report/3.Theory/Parser.tex

```diff
@@ -0,0 +1,112 @@
+The major part of the contextual analysis is the type checking.
+
+\textbf{Type checking}
+
+The type checking part of the parser often utilizes a symbol table.
+
+INSERT JONAS SYMBOL TABLE STUFF HERE.
+
+It is also the responsibility of this part of the parser to provide any relevant type information that may be needed in order to proceed with compilation in later phases. 
+There are several ways to inform the later stages of the compiler about the needed type information. 
+One solution is to store and pass the symbol table to the later stages of the compiler. 
+
+Another way is to decorate the abstract syntax tree with the needed type information. 
+This is called a decorated syntax tree, and will in most cases be the final output of the parser.
+An example of decorating an abstract syntax tree could be when using implicit casts. 
+Most languages have this feature between integers and floating point numbers. 
+So if the user were to do arithmetic with an integer and a floating point number, the integer would then be implicitly casted to a floating point number.
+
+\subsection{Single parse and multi parse}
+
+Depending on the language, a single run through the passer may not be enough to validate the input or generate the output. 
+For instance if the language allows calling of functions or usage of classes that has not yet been declared. 
+In this case the parser should do a single parse to map out the functions and classes that can be used. 
+After that another parse should be done, where the information from the previous parse is utilized to ensure the validity of the input. 
+
+Surely single pass parsers are the fastest. 
+However it does limit the language a bit, and therefore most mainstream compilers are multi parse compilers these days.
+
+\subsection{Parser types}
+
+Parsers comes in a wide variety of different flavors. 
+Generally there are two different approaches to creating a parser.
+
+\textbf{Top down parsing}
+
+LL parsers uses the top down parsing strategy. 
+LL means that the parser is Left to right using Left most derivation. 
+If a CFG is LL(K) it means that the parser needs to do at max K look aheads in order to parse the sentence. 
+LL parsers can be implemented as a table parser, or also often as a recursive decent parser. 
+
+\textbf{Bottom up parsing}
+
+There are a couple of different bottom up parsers. 
+Namely the LR(K) and the LALR parsers are interesting. 
+The bottom up parsers can recognize a lot more languages than the top down parsers.
+However, they are also generally more complex to implement.
+Because of this bottom up parsers are only rarely handmade.
+They will usually be implemented using the shift reduce strategy.
+
+
+\subsection{Parser tools}
+
+Like for the lexer, a vast amount of tools exist in order to ease the parsing part of the compiler. 
```

Comment:
"Just as with lexers" is stedet for "Like for the lexer"

eller noget andet, bare det lyder bedre